### DISCUSSION TITLE
Inverse problem for a ODE system with exogenous input in an integral

### DISCUSSION THREAD

**Discussion by aoguedao** [lululxvi/deepxde#1162] [Q&A]
Hello everyone  - 

I'm trying to estimate a few parameters in a ODE system but the setup this time is a little different since it needs an integral in one of the equations. I have been reading FAQs, examples, documentation and even source code but I'm not sure how to deal with it. 

In order to explain it I will use a Lorentz system since it is similar to [this example](https://deepxde.readthedocs.io/en/latest/demos/pinn_inverse/lorenz.inverse.forced.html).

$$\displaylines{ \frac{dx}{dt} = \sigma(y-x) + \int_0^t q(t - \eta) e^{-a \eta} d\eta \\\
\frac{dy}{dt} = x (\rho - z) - y \\\
\frac{dz}{dt} = x y - \beta z}$$

where $q(\eta)$ is a function I know its values and $a$ is a parameter to be learned.

As far as I know IDE current version only supports 1D problems :( 

I would really appreciate any insight, I was thinking to use `auxiliary_var_function` argument but the integral $\int_0^t q(t - \eta) e^{-a \eta} d\eta$ depends of a learnable parameter. 

I'm willing to add this new functionality if it is needed and even I could make a pull-request in the future. 

Thanks in advance

**Comment by lululxvi** [comment]
A simple solution is to expand the integral into a sum.

**Reply by aoguedao** [reply]
Thank you so much for your answer, I have been trying to different approaches but I haven't been able to expand the integral into a sum. When I build synthetic data I had no issue using scipy

```{python}
def integral(t, a):
    dt = 0.1
    nu_array = np.arange(dt, t + dt, dt)
    value = np.sum(dphi_dt(t - nu_array, b, c) * dv_dt(nu_array, d)) * dx
    return value
```

However, when I try to do the same with `tensorflow.compat.v1` backend 

```{python}
def tf_integral(t_tensor, a):
    dt = tf.constant(0.1)
    ta = tf.TensorArray(tf.float32, size=0, dynamic_size=True)
    ta = ta.unstack([0., 1.])
    for i in range(n_days):
        t = t_tensor[i]
        # breakpoint()
        nu_tensor = tf.range(dt, t + dt, dt)
        value = tf.reduce_sum(tf_dphi_dt(t - nu_tensor, b, c) * tf_dv_dt(nu_tensor, d)) * dx
        ta = ta.write(i, value)
    return ta.stack()
```
where `n_days` is a integer since I wasn't able to get something like `len(t_tensor)`.  But I got the following error.

```
ValueError: in user code:

    File "/mnt/d/Documents/git/insight/notebooks/china project/china-v-fitted.py", line 130, in tf_integral_dv_dphi  *
        nu_tensor = tf.range(dx, t + dx, dx)

    ValueError: Shape must be rank 0 but is rank 1
         for 'limit' for '{{node range}} = Range[Tidx=DT_FLOAT](Const, add, Const)' with input shapes: [], [1], [].
```

I'm not sure if I can use  `@tf.function` decorator  with `tensorflow.compat.v1` backend since it is my first time working with tensorflow.

I've been looking for alternatives and it seems I could use [tf.compat.v1.while_loop](https://www.tensorflow.org/api_docs/python/tf/compat/v1/while_loop) and/or [tf.compat.v1.wrap_function](https://www.tensorflow.org/api_docs/python/tf/compat/v1/wrap_function) but honestly I'm not sure.

**Reply by lululxvi** [reply]
See the example code of IDE https://github.com/lululxvi/deepxde/blob/master/deepxde/data/ide.py
